### PR TITLE
Update controller-runtime

### PR DESCRIPTION
**Pull request by dergeberl**
- Update go version to 1.20
- Update controller-runtime, controller-tools

- Add tests that run on multible k8s versions

### CODE REVIEW

File: go.mod [hunk 1]
```diff
@@ -20,62 +20,47 @@ require (
...
-replace (
-	k8s.io/api => k8s.io/api v0.24.12
-	k8s.io/apimachinery => k8s.io/apimachinery v0.24.12
-	k8s.io/client-go => k8s.io/client-go v0.24.12
+	k8s.io/api v0.27.3
+	k8s.io/apimachinery v0.27.3
+	k8s.io/client-go v0.27.3
+	k8s.io/utils v0.0.0-20230505201702-9f6742963106
+	sigs.k8s.io/controller-runtime v0.15.0
+	sigs.k8s.io/controller-runtime/tools/setup-envtest v0.0.0-20230702162849-a23f3603e1dc
```
Comment (timebertt) [inline go.mod:28]: Optionally, pin this to the same version as controller-runtime.
```suggestion
sigs.k8s.io/controller-runtime/tools/setup-envtest v0.0.0-20230523032821-116a1b831fff // v0.15.0
```

File: cmd/yawol-cloud-controller/main.go [hunk 2]
```diff
@@ -138,10 +139,12 @@ func main() {
...
-		Scheme:                        scheme,
-		MetricsBindAddress:            "0",
-		Port:                          9443,
-		Namespace:                     *infrastructureDefaults.Namespace,
+		Scheme:             scheme,
+		MetricsBindAddress: "0",
+		NewCache: func(config *rest.Config, opts cache.Options) (cache.Cache, error) {
+			opts.Namespaces = append(opts.Namespaces, *infrastructureDefaults.Namespace)
+			return cache.New(config, opts)
+		},
```
Comment (timebertt) [inline cmd/yawol-cloud-controller/main.go:147]: `NewCache` is a low-level primitive but not required here. You can use the cache options:
```suggestion
		Cache: cache.Options{
			Namespaces: []string{*infrastructureDefaults.Namespace},
		},
```

File: cmd/yawol-controller/main.go [hunk 3]
```diff
@@ -153,15 +155,17 @@ func main() {
...
 			LeaderElectionID:              "3a7ac996.stackit.cloud",
 			LeaseDuration:                 &leasesDuration,
 			RenewDeadline:                 &leasesRenewDeadline,
 			RetryPeriod:                   &leasesRetryPeriod,
 			LeaderElectionResourceLock:    leasesLeaderElectionResourceLock,
-			Namespace:                     clusterNamespace,
+			NewCache: func(config *rest.Config, opts cache.Options) (cache.Cache, error) {
+				opts.Namespaces = append(opts.Namespaces, clusterNamespace)
+				return cache.New(config, opts)
+			},
```
Comment (timebertt) [inline cmd/yawol-controller/main.go:168]: Same here
```suggestion
		Cache: cache.Options{
			Namespaces: []string{clusterNamespace},
		},
```

File: cmd/yawol-controller/main.go [hunk 4]
```diff
@@ -189,15 +193,17 @@ func main() {
...
 			LeaderElectionID:              "rgp5vg43.stackit.cloud",
 			LeaseDuration:                 &leasesDuration,
 			RenewDeadline:                 &leasesRenewDeadline,
 			RetryPeriod:                   &leasesRetryPeriod,
 			LeaderElectionResourceLock:    leasesLeaderElectionResourceLock,
-			Namespace:                     clusterNamespace,
+			NewCache: func(config *rest.Config, opts cache.Options) (cache.Cache, error) {
+				opts.Namespaces = append(opts.Namespaces, clusterNamespace)
+				return cache.New(config, opts)
+			},
```
Comment (timebertt) [inline cmd/yawol-controller/main.go:206]: Same here
```suggestion
		Cache: cache.Options{
			Namespaces: []string{clusterNamespace},
		},
```

File: cmd/yawol-controller/main.go [hunk 5]
```diff
@@ -240,15 +246,17 @@ func main() {
...
 			LeaderElectionID:              "tanf7ges.stackit.cloud",
 			LeaseDuration:                 &leasesDuration,
 			RenewDeadline:                 &leasesRenewDeadline,
 			RetryPeriod:                   &leasesRetryPeriod,
 			LeaderElectionResourceLock:    leasesLeaderElectionResourceLock,
-			Namespace:                     clusterNamespace,
+			NewCache: func(config *rest.Config, opts cache.Options) (cache.Cache, error) {
+				opts.Namespaces = append(opts.Namespaces, clusterNamespace)
+				return cache.New(config, opts)
+			},
```
Comment (timebertt) [inline cmd/yawol-controller/main.go:259]: Same here
```suggestion
		Cache: cache.Options{
			Namespaces: []string{clusterNamespace},
		},
```

File: cmd/yawollet/main.go [hunk 6]
```diff
@@ -187,18 +188,12 @@ func main() {
 	mgr, err := ctrl.NewManager(ctrl.GetConfigOrDie(), ctrl.Options{
 		Scheme:             scheme,
 		MetricsBindAddress: metricsAddr,
-		Port:               9443,
 		LeaderElection:     false,
-		Namespace:          namespace,
-
 		NewCache: func(config *rest.Config, opts cache.Options) (cache.Cache, error) {
```
Comment (timebertt) [inline cmd/yawollet/main.go:192]:
```suggestion
		Cache: cache.Options{
			Namespaces: []string{namespace},
			ByObject: map[client.Object]cache.ByObject{
				&yawolv1beta1.LoadBalancer{}:        {Field: fields.SelectorFromSet(fields.Set{"metadata.name": loadbalancerName})},
				&yawolv1beta1.LoadBalancerMachine{}: {Field: fields.SelectorFromSet(fields.Set{"metadata.name": loadbalancerMachineName})},
			},
		},
```

File: controllers/yawol-controller/loadbalancer/loadbalancer_controller.go [hunk 7]
```diff
@@ -139,8 +138,9 @@ func (r *Reconciler) SetupWithManager(mgr ctrl.Manager) error {
 		For(&yawolv1beta1.LoadBalancer{}).
 		// we can't use Owns because it filters for ownerReference.controller==true which is not the case for our objects (as of now)
 		Watches(
```
Comment (timebertt) [inline controllers/yawol-controller/loadbalancer/loadbalancer_controller.go:140]: With the new version, we can use `Owns` again:
```go
		Owns(&yawolv1beta1.LoadBalancerSet{}, builder.MatchEveryOwner, builder.WithPredicates(LoadBalancerSetPredicate())).
```

**Review by timebertt — COMMENTED** [review]
Nice, some suggestions for simplification inline.

**Review by einfachnuralex — APPROVED** [review]

**Review by timebertt — APPROVED** [review]
Awesome!

**Pull request merged**